### PR TITLE
Dev env

### PR DESCRIPTION
**Pull request by BabaBert**
Fixes https://github.com/qdrant/qdrant/issues/2117

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?

There are two Python tests, that fail within `openapi_integration/test_snapshot.py`: `test_snapshot_operations`. Since I haven't made changes to the code or the tests, I can't verify, if the issues are related to the branch or are related to the dev-envionment.

~~Have you formatted your code locally using `cargo fmt` command prior to submission?~~
~~Have you checked your code using `cargo clippy` command?~~

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

Added a `devcontainer`-configuration based on the existing Dockerfile to ensure, that the tests execute correctly locally.

* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

I downgraded the debian-image to version 11 to avoid https://github.com/devcontainers/features/issues/598

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -88,11 +88,14 @@ RUN PATH="$PATH:/opt/mold/bin" \
...
 
+# to skip the compilation for testing
+# RUN mkdir /qdrant/qdrant
+
 
 # Download and extract web UI
 RUN mkdir /static ; STATIC_DIR='/static' ./tools/sync-web-ui.sh
 
-FROM debian:12-slim AS qdrant
+FROM debian:11-slim AS qdrant
```
Comment (generall) [inline Dockerfile:98]: is it just a merge artifact or intentional?

Comment (BabaBert) [reply Dockerfile:98]: Intentional because of https://github.com/devcontainers/features/issues/598. I have to see if there are other workarounds if this is a dealbreaker. This would be only temporary until the docker-in-docker feature is patched.

**Review by coszio — CHANGES_REQUESTED** [review]
It's true that development is very Linux centric. Having something like this will indeed make it easier for people working on other platforms to contribute. However, I see a few downsides for this approach:
1. build times become ridiculously long. Cold Debug build took about 2 hours, when normally it takes less than 10 mins. My intuition says it's because it's syncing the local files with the ones in docker, similar to [this](https://docs.docker.com/desktop/dev-environments/create-dev-env/#launch-from-a-local-folder)
2. Keeping the configuration in source control means enabling it for everyone, such that these kind of messages appear when opening the repo without containerizing it.

    imagen

3. It's more or less tied with a single editor (VSCode)
 
Another solution is to include the steps to set up a certain dev environment within the existing `DEVELOPMENT.md` file. Currently we only have the Linux/Debian section. For Mac the instructions would be pretty similar by using `brew`. For Windows I suspect it should be possible using WSL.

Lastly, I am requesting changes because there are some changes to proto files which seem to belong to another PR 😄

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        "dockerfile": "../Dockerfile"
+    },
+    "features": {
+        "ghcr.io/devcontainers/features/common-utils:latest": {},
+		"ghcr.io/devcontainers/features/docker-in-docker:latest": {},
+		"ghcr.io/devcontainers/features/rust:latest": {},
+        "ghcr.io/devcontainers-contrib/features/protoc:latest": {} 
+    },
+    "postCreateCommand": "apt-get update && apt-get install -y clang lld cmake protobuf-compiler jq && rustup component add rustfmt"
+}
```
Comment (coszio) [inline .devcontainer/devcontainer.json:16]:
```suggestion
{
  "name": "Debian",
  "image": "mcr.microsoft.com/devcontainers/base:bullseye",

  "features": {
    "ghcr.io/devcontainers/features/common-utils:latest": {},
    "ghcr.io/devcontainers/features/docker-in-docker:latest": {},
    "ghcr.io/devcontainers/features/rust:latest": {
      "profile": "default"
    },
    "ghcr.io/devcontainers-contrib/features/protoc:latest": {}
  },
  "postCreateCommand": {
    "apt-get": "sudo apt-get update -y && sudo apt-get upgrade -y && sudo apt-get install -y clang lld cmake protobuf-compiler jq gcc g++",
    // mostly for running cargo +nightly fmt
    "nightly": "rustup toolchain install nightly"
  },

  // Use 'forwardPorts' to make a list of ports inside the container available locally.
  "forwardPorts": [6333, 6334]
}

```

**Review by coszio — COMMENTED** [review]
I think we can avoid using and modifying the existing Dockerfile, as it is intended only for compiling production releases, which might explain why it takes so long to build a debug profile. 
Tested the suggested change locally (macOS) and in codespaces and it builds and runs tests fine.

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (BabaBert) [reply .devcontainer/devcontainer.json:16]: > I think we can avoid using and modifying the existing Dockerfile, as it is intended only for compiling production releases, which might explain why it takes so long to build a debug profile.

Yes, definitely, I was just about to make a commit with a separate Dockerfile:

```Dockerfile

ARG RUST_VERSION

FROM rust:${RUST_VERSION:-1}-bullseye

RUN apt-get update && \
    apt-get install -y ca-certificates tzdata \
    clang lld cmake protobuf-compiler jq 

RUN rustup toolchain install nightly && \
    rustup component add rustfmt --toolchain nightly

ARG APP=/qdrant

RUN mkdir -p ${APP}

WORKDIR ${APP}

EXPOSE 6333
EXPOSE 6334

CMD ["./entrypoint.sh"]
```

I think the resulting differences are marginal. Going with the JSON is probably easier to grasp.

File: Dockerfile (hunk 1 above)
Comment (coszio) [reply Dockerfile:98]: We can ditch these changes now that we are relying on the `devcontainer.json` only

**Review by coszio — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -88,6 +88,9 @@ RUN PATH="$PATH:/opt/mold/bin" \
     && PROFILE_DIR=$(if [ "$PROFILE" = dev ]; then echo debug; else echo $PROFILE; fi) \
     && mv target/$(xx-cargo --print-target-triple)/$PROFILE_DIR/qdrant /qdrant/qdrant
 
+# to skip the compilation for testing
+# RUN mkdir /qdrant/qdrant
+
```
Comment (coszio) [inline Dockerfile:93]:
```suggestion
```

File: compose-dev.yaml
Comment (coszio) [inline compose-dev.yaml:1]: Docker's dev environments is still [in Beta](https://docs.docker.com/desktop/dev-environments/), and **is** expected to change soon(-ish?) so I think this should go in another PR, favoring devcontainer solution here.

**Review by coszio — CHANGES_REQUESTED** [review]

File: compose-dev.yaml
Comment (BabaBert) [reply compose-dev.yaml:1]: Alright. Though, that should solve for the performance issues and the choice for the IDE going forward.

**Comment by BabaBert** [comment]
Should be ready to merge now.

**Review by coszio — APPROVED** [review]
Thanks for pushing this forward, I'll use it myself in the near future

**Comment by coszio** [comment]
After discussion, we agreed that this config should **not** be committed to the repo as it is right now (see #2330). 
However, I do feel we should not just close this PR, so I propose to add it as optional local setup instead.

**Comment by IvanPleshkov** [comment]
@BabaBert dev containers are a great feature. However, we will not merge this PR because repo does not contain IDE/env-related files like `.vscode` or `.idea` even though these files contain a debugging setup. Also, there are problems with raft that you described recently in another PR.
dev containers are already included in `.gitignore` so you can continue using dev containers for personal purposes: https://github.com/qdrant/qdrant/pull/2330

**Comment by BabaBert** [comment]
will do

**Comment by BabaBert** [comment]
https://gist.github.com/BabaBert/8fa83d51431be28babe16db551edcf91
